moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp

```diff
@@ -35,167 +35,173 @@
-
-        if (controller_list[i]["joints"].getType() != XmlRpc::XmlRpcValue::TypeArray)
+        std::vector<std::string> controller_joints;
+        if (!node_->get_parameter(PARAM_BASE_NAME + "." + controller_name + ".joints", controller_joints))
         {
-          ROS_ERROR_STREAM_NAMED("MoveItFakeControllerManager", "The list of joints for controller "
-                                                                    << name << " is not specified as an array");
+          RCLCPP_ERROR_STREAM(LOGGER, "No joints specified for controller " << controller_name);
           continue;
         }
-        std::vector<std::string> joints;
-        joints.reserve(controller_list[i]["joints"].size());
-        for (int j = 0; j < controller_list[i]["joints"].size(); ++j)
-          joints.emplace_back(std::string(controller_list[i]["joints"][j]));
 
-        const std::string& type =
-            controller_list[i].hasMember("type") ? std::string(controller_list[i]["type"]) : DEFAULT_TYPE;
+        std::string type = DEFAULT_TYPE;
+        if (node_->has_parameter(PARAM_BASE_NAME + "." + controller_name + ".type"))
+          node_->get_parameter(PARAM_BASE_NAME + "." + controller_name + ".type", type);
+
         if (type == "last point")
-          controllers_[name].reset(new LastPointController(name, joints, pub_));
+          controllers_[controller_name].reset(new LastPointController(controller_name, controller_joints, pub_));
         else if (type == "via points")
-          controllers_[name].reset(new ViaPointController(name, joints, pub_));
+          controllers_[controller_name].reset(new ViaPointController(controller_name, controller_joints, pub_));
         else if (type == "interpolate")
-          controllers_[name].reset(new InterpolatingController(name, joints, pub_));
+        {
+          //@note: would put this in InterpolatingController's constructor
+          // but they disabled WallRate::operator= for some reason and that complicates things
+          double rate = 10.0;
+          if (node_->has_parameter("fake_interpolating_controller_rate"))
+            node_->get_parameter("fake_interpolating_controller_rate", rate);
+          controllers_[controller_name].reset(
+              new InterpolatingController(controller_name, controller_joints, pub_, rate));
+        }
         else
-          ROS_ERROR_STREAM("Unknown fake controller type: " << type);
+          RCLCPP_ERROR_STREAM(LOGGER, "Unknown fake controller type: " << type);
       }
       catch (...)
       {
-        ROS_ERROR_NAMED("MoveItFakeControllerManager", "Caught unknown exception while parsing controller information");
+        RCLCPP_ERROR(LOGGER, "Caught unknown exception while parsing controller information");
       }
     }
   }
 
-  sensor_msgs::JointState loadInitialJointValues(XmlRpc::XmlRpcValue& param) const
-  {
-    sensor_msgs::JointState js;
+  // TODO: codebase wide refactoring for XmlRpc
```

Comment:
We could easily work around missing XmlRpc support by structuring the parameters like this:

```
initial:
  <group>: <pose>
```

Since we know the robot model, it would be straight forward to just look for parameters using the group names and then just use the existing checks if `pose` is a valid named target. We don't need an array of structs if we already know that group names are unique identifiers and can be used as parameter keys.

Reply:
ported e9b3c1d